comfy/caching.py

```diff
@@ -0,0 +1,299 @@
+import itertools
+from typing import Sequence, Mapping
+from comfy.graph import DynamicPrompt
+
+import nodes
+
+from comfy.graph_utils import is_link
+
+class CacheKeySet:
+    def __init__(self, dynprompt, node_ids, is_changed_cache):
+        self.keys = {}
+        self.subcache_keys = {}
+
+    def add_keys(self, node_ids):
+        raise NotImplementedError()
+
+    def all_node_ids(self):
+        return set(self.keys.keys())
+
+    def get_used_keys(self):
+        return self.keys.values()
+
+    def get_used_subcache_keys(self):
+        return self.subcache_keys.values()
+
+    def get_data_key(self, node_id):
+        return self.keys.get(node_id, None)
+
+    def get_subcache_key(self, node_id):
+        return self.subcache_keys.get(node_id, None)
+
+class Unhashable:
+    def __init__(self):
+        self.value = float("NaN")
+
+def to_hashable(obj):
+    # So that we don't infinitely recurse since frozenset and tuples
+    # are Sequences.
+    if isinstance(obj, (int, float, str, bool, type(None))):
+        return obj
+    elif isinstance(obj, Mapping):
+        return frozenset([(to_hashable(k), to_hashable(v)) for k, v in sorted(obj.items())])
+    elif isinstance(obj, Sequence):
+        return frozenset(zip(itertools.count(), [to_hashable(i) for i in obj]))
+    else:
+        # TODO - Support other objects like tensors?
```

Comment:
can probably use `.id()` or `data_ptr()` or something for a quick good-enough hash source if tensors are needed here

Reply:
I did some investigation on this before giving up on tensors (for now). Unfortunately, due to torch's use of wrappers (and the fact that the 'tensor' you received may just be a wrapper around an existing tensor), it didn't look like either `.id()` or `.data_ptr()` were reliably unique or constant: https://discuss.pytorch.org/t/why-did-i-get-the-same-two-ids-when-using-id-function-on-two-different-indexes-of-a-pytorch-tensor/150469

Additionally, because only constant inputs are part of the cache key (and the HTTP endpoint doesn't support specifying constant tensors), this is only an issue if a custom node uses a tensor constant during node expansion (which is an edge-case solvable by passing a link instead).